### PR TITLE
General: Workfiles builder using query functions

### PR DESCRIPTION
**Pull request by iLLiCiTiT**
## Brief description
Initial refactor of new workfile template builder which avoids using direct mongo calls. It is not backwards compatible change but result should be the same.

## Description
Function `get_representations` can filter representations by context keys even with regexes. Replaced `convert_to_db_filters` method with `get_representations` so placeholder returns already queried representations instead of returning filters. Maya's placeholder has implemented `get_representations` which should work as before but using `get_representations` function.

## Additiona info
Made few smaller changes in abstract placeholder class. Added few properties and changed few names of properties to be more specific. Pass full asset documents instead of passing just their names (they're available anyway). Required and stored data did not change so it should be backwards compatibile for existing templates.

## Testing notes:
- [ ] Deadline should find representation resources as did (do we need that functionality?)
- [ ] Placeholder templates should work as they did in Maya

### CODE REVIEW

**Review by m-u-r-p-h-y — CHANGES_REQUESTED** [review]
Maya is starting without OP menu

```
D:\REPO\OpenPype\.venv\lib\site-packages\qtpy\__init__.py:230: RuntimeWarning: Selected binding "pyqt5" could not be found, using "pyside2"
  'using "{}"'.format(initial_api, API), RuntimeWarning)
Failed to execute userSetup.py
Traceback (most recent call last):
  File "D:\REPO\OpenPype\.venv\lib\site-packages\past\builtins\misc.py", line 87, in execfile
    exec_(code, myglobals, mylocals)
  File "D:/REPO/OpenPype/openpype/hosts/maya/startup/userSetup.py", line 4, in <module>
    from openpype.hosts.maya.api import MayaHost
  File "D:\REPO\OpenPype\openpype\hosts\maya\api\__init__.py", line 7, in <module>
    from .pipeline import (
  File "D:\REPO\OpenPype\openpype\hosts\maya\api\pipeline.py", line 32, in <module>
    from . import menu, lib
  File "D:\REPO\OpenPype\openpype\hosts\maya\api\menu.py", line 13, in <module>
    from openpype.pipeline.workfile.build_template import (
  File "D:\REPO\OpenPype\openpype\pipeline\workfile\build_template.py", line 6, in <module>
    from .abstract_template_loader import (
  File "D:\REPO\OpenPype\openpype\pipeline\workfile\abstract_template_loader.py", line 234
    [placeholder.name] for placeholder in placeholders]
                                                      ^
SyntaxError: invalid syntax
```

**Comment by iLLiCiTiT** [comment]
> Maya is starting without OP menu

Sorry I didn't resolve merge conflict right way.

**Comment by ClementHector** [comment]
I can check with pleasure!

**Review by m-u-r-p-h-y — CHANGES_REQUESTED** [review]
There are two errors popping up during publish:

Collect Versions Loaded in Scene

```
Traceback (most recent call last):
  File "D:\REPO\OpenPype\.venv\lib\site-packages\pyblish\plugin.py", line 522, in __explicit_process
    runner(*args)
  File "D:\REPO\OpenPype\openpype\plugins\publish\collect_scene_loaded_versions.py", line 51, in process
  File "D:\REPO\OpenPype\openpype\client\entities.py", line 1219, in get_representations
    extensions=extensions,
NameError: name 'extensions' is not defined
```


Collect Render Layers

```
Message: 
'RenderProductsRedshift' object has no attribute 'aov_separator'
Line: 
966
Traceback: 
Traceback (most recent call last):
  File "D:\REPO\OpenPype\.venv\lib\site-packages\pyblish\plugin.py", line 522, in __explicit_process
    runner(*args)
  File "D:\REPO\OpenPype\openpype\hosts\maya\plugins\publish\collect_render.py", line 165, in process
  File "D:\REPO\OpenPype\openpype\hosts\maya\api\lib_renderproducts.py", line 181, in get
    return renderer(layer, render_instance)
  File "D:\REPO\OpenPype\openpype\hosts\maya\api\lib_renderproducts.py", line 202, in __init__
    self.layer_data = self._get_layer_data()
  File "D:\REPO\OpenPype\openpype\hosts\maya\api\lib_renderproducts.py", line 321, in _get_layer_data
    file_prefix = self.get_renderer_prefix()
  File "D:\REPO\OpenPype\openpype\hosts\maya\api\lib_renderproducts.py", line 966, in get_renderer_prefix
    prefix = "{}{}<aov>".format(prefix, self.aov_separator)
AttributeError: 'RenderProductsRedshift' object has no attribute 'aov_separator'
```

**Comment by iLLiCiTiT** [comment]
> Collect Versions Loaded in Scene

Fixed

> Collect Render Layers

I'm afraid that is not caused by this PR. Can you try it in develop?

**Comment by iLLiCiTiT** [comment]
@ClementHector what is the `node` argument in placeholder? Is it a string or object and what it represents? Considering it will not be a node all the time it should be renamed to something universal. Could that be renamed to `placeholder_identifier` (or similar)?

**Comment by ClementHector** [comment]
Node is a string, it allows us to find what was imported in order to be able to make updates. it is very similar to the `avalon container`. Can you call it `placeholder_identifier` or `placeholder_container`?

**Review by m-u-r-p-h-y — COMMENTED** [review]
```
# Error: $and/$or/$nor must be a nonempty array, full error: {'ok': 0.0, 'errmsg': '$and/$or/$nor must be a nonempty array', 'code': 2, 'codeName': 'BadValue'}
# Traceback (most recent call last):
#   File "D:\REPO\OpenPype\openpype\pipeline\workfile\build_template.py", line 30, in build_workfile_template
#     template_loader.populate_template()
#   File "D:\REPO\OpenPype\openpype\pipeline\workfile\abstract_template_loader.py", line 272, in populate_template
#     for representation in placeholder_representations:
#   File "D:\REPO\OpenPype\openpype\pipeline\workfile\abstract_template_loader.py", line 304, in get_placeholder_representations
#     linked_asset_docs
#   File "D:\REPO\OpenPype\openpype\hosts\maya\api\template_loader.py", line 238, in get_representations
#     context_filters=context_filters
#   File "D:\REPO\OpenPype\.venv\lib\site-packages\pymongo\cursor.py", line 1238, in next
#     if len(self.__data) or self._refresh():
#   File "D:\REPO\OpenPype\.venv\lib\site-packages\pymongo\cursor.py", line 1155, in _refresh
#     self.__send_message(q)
#   File "D:\REPO\OpenPype\.venv\lib\site-packages\pymongo\cursor.py", line 1045, in __send_message
#     operation, self._unpack_response, address=self.__address)
#   File "D:\REPO\OpenPype\.venv\lib\site-packages\pymongo\mongo_client.py", line 1426, in _run_operation
#     address=address, retryable=isinstance(operation, message._Query))
#   File "D:\REPO\OpenPype\.venv\lib\site-packages\pymongo\mongo_client.py", line 1525, in _retryable_read
#     return func(session, server, sock_info, secondary_ok)
#   File "D:\REPO\OpenPype\.venv\lib\site-packages\pymongo\mongo_client.py", line 1422, in _cmd
#     unpack_res)
#   File "D:\REPO\OpenPype\.venv\lib\site-packages\pymongo\server.py", line 130, in run_operation
#     _check_command_response(first, sock_info.max_wire_version)
#   File "D:\REPO\OpenPype\.venv\lib\site-packages\pymongo\helpers.py", line 167, in _check_command_response
#     raise OperationFailure(errmsg, code, response, max_wire_version)
# pymongo.errors.OperationFailure: $and/$or/$nor must be a nonempty array, full error: {'ok': 0.0, 'errmsg': '$and/$or/$nor must be a nonempty array', 'code': 2, 'codeName': 'BadValue'} # 
```

**Review by m-u-r-p-h-y — APPROVED** [review]
now the build Workfile from template works:

```
# # >>> [ BUILDING ASSET FROM TEMPLATE :
# Starting templated build for testCube in murphy_22101
# 
# Asset : testCube
# Task : modeling (Modeling)
# Host : maya
# Project : murphy_22101
 ] 
# openpype.lib.profiles_filtering : Looking for matching profile for: task_types: "Modeling" | tasks: "modeling" # 
# openpype.lib.profiles_filtering : Profile selected: {'task_types': ['Modeling'], 'tasks': [], 'path': '{STUDIO_CFG}\\maya\\templates\\template_modeling_v001.ma'} # 
# >>> [ Found template at: 'D:\REPO\STUDIO\maya\templates\template_modeling_v001.ma' ] 
# File read in  0.014 seconds. # 
#   - { < MayaPlaceholder |review_mayascene_assets > }: [ Valid placeholder : |review_mayascene_assets ] 
#   - { < MayaPlaceholder |Linked_model_asset > }: [ Valid placeholder : |Linked_model_asset ] 
#   - { MayaTemplateLoader }: [ Placeholders found in template: ['|review_mayascene_assets', '|Linked_model_asset'] ] 
#   - { MayaTemplateLoader }: [ Start to processing placeholder |review_mayascene_assets ] 
#   - { MayaTemplateLoader }: [ Start to processing placeholder |Linked_model_asset ] 
# # >>> [ Loading bob_modelMain with loader ReferenceLoader
Loader arguments used :  ] 
# openpype.pipeline.load.utils : Running 'ReferenceLoader' on 'bob' # 
# SubsetLoader : {} # 
# File read in  0.044 seconds. # 
```

File: openpype/pipeline/workfile/abstract_template_loader.py [hunk 1]
```diff
@@ -377,107 +400,144 @@ def get_template_nodes(self):
...
 
     @property
     def is_context(self):
-        """Return placeholder type
+        """Check if is placeholder context type.
+
         context_asset: For loading current asset
         linked_asset: For loading linked assets
+
+        Question:
```
Comment (ClementHector) [inline openpype/pipeline/workfile/abstract_template_loader.py:475]: This property can be deleted.

File: openpype/pipeline/workfile/abstract_template_loader.py [hunk 2]
```diff
@@ -377,107 +400,144 @@ def get_template_nodes(self):
...
-            Int: Order priority
-        """
-        return self.data.get('order')
+        return self.data["order"]
 
     @property
-    def loader(self):
-        """Return placeholder loader type
+    def loader_name(self):
+        """Return placeholder loader type.
```
Comment (ClementHector) [inline openpype/pipeline/workfile/abstract_template_loader.py:459]: """Return placeholder loader name.

**Pull request merged**